### PR TITLE
Remote ota/development

### PR DESCRIPTION
**Pull request by chezlyne**

### CODE REVIEW

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 1]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+            }
+
+            return false;
+        }
+        else
+        {
+            rsp = &m_responses[m_nextCmd];
+    
+            /*Update the command with required content to write in the flash memory.*/
+            if(true == updateFwProgCmd(m_updatedCmd))
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:187]: add blank after if

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 2]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+        else if (m_nextCmd == FLSPRG_WRITE_BLOCK)
+        {
+            /*Update the command identifier bytes for write block*/
+            m_updatedCmdBuffer[0] = 0x42;
+            m_updatedCmdBuffer[1] = 0x00;
+            m_updatedCmdBuffer[2] = 0x80;
+            m_updatedCmdBuffer[3] = 0x46;
+            /*Update the command for WRITE_MEMORY_PAGE*/
+            m_firmwareBytesRead = m_firmwareFile.read(
+                &m_firmwareBuffer[WRITE_BLOCK_CMD_LENGTH],
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:289]: BUG: Must be m_updatedCmdBuffer instead

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 3]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+                return false;
+            }
+            /*If the number of bytes read is less than MAX_BYTES_TO_READ,fill the rest of the buffer with a pattern.*/
+            size_t gapFill = FLASH_BLOCK_LENGTH - m_firmwareBytesRead;
+            if (0U != gapFill)
+            {
+                m_nextCmd = FLSPRG_COMPLETE;
+                while (gapFill > 0U)
+                {
+                    m_firmwareBuffer[WRITE_BLOCK_CMD_LENGTH + m_firmwareBytesRead] = 0xFF;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:310]: m_updatedCmdBuffer

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 4]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+     */
+    static const size_t WRITE_BLOCK_CMD_LENGTH = 4U;
+
+    static  CommandInfo m_cmds[]; /**< Array of command information.*/
+    static  ResponseInfo m_responses[]; /**< Array of response information.*/
+    
+    CommandInfo m_updatedCmd; /**<Intermediate storage for updated commands.*/
+
+    const char* m_fileName = "/firmware.bin"; /**< Name of the file to be read.*/
+    uint8_t m_firmwareBuffer[128]; /**< Buffer to hold firmware data read from the file.*/
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:222]: remove this one,  we only use m_updatedCmdBuffer

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 5]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+
+    /**
+     * @brief Enumeration of programming commands.
+     * This enumeration defines the various programming commands used during firmware programming.
+     * It includes commands for setting memory addresses and writing data blocks.
+     */
+   enum ProgCmds {
+    FLSPRG_SET_ADDR,    /**< Command to set the memory address. */
+    FLSPRG_WRITE_BLOCK,  /**< Command to write a data block to memory.*/
+    FLSPRG_COMPLETE   /**<Signalize the end of the Flashing.*/
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:234]: indent comments

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 6]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+    bool updateFwProgCmd(CommandInfo& fwProgCmd)
+    {
+        if (m_nextCmd == FLSPRG_SET_ADDR)
+        {
+            /*Perform the required updates.*/
+            m_updatedCmdBuffer[0] = 0x41;  /**< The first value is always 0x41.*/
+            m_updatedCmdBuffer[1] = (m_currWriteMemAddr >> 8) & 0xFF; /**<Higher 8 bits of m_currWriteMemAddr.*/
+            m_updatedCmdBuffer[2] = m_currWriteMemAddr & 0xFF; //**<Lower 8 bits of m_currWriteMemAddr.*/
+            /*Increase m_currWriteMemAddr by 64 for the next command.*/
+            m_currWriteMemAddr += 64U;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:273]: 64 ? I thought it is FLASH_BLOCK_LENGTH (128)

File: lib/ZumoOta/src/FileManager.h [hunk 7]
```diff
@@ -15,7 +15,8 @@
 /******************************************************************************
  * Includes
  *****************************************************************************/
-
+ #include <cstdint>
```
Comment (nhjschulz) [inline lib/ZumoOta/src/FileManager.h:18]: I don't see why you need these 2 includes here.

File: lib/ZumoOta/src/FlashManager.cpp [hunk 8]
```diff
@@ -108,37 +108,29 @@ size_t FlashManager::readingStream(uint8_t* expectedResponse, size_t mybytes)
...
-       else
-       {
-       
         /* Send the OpCode and command data to Zumo robot. */
         size_t bytesWritten= deviceStream.write(command, commandsize);
-      
-         LOG_INFO("byteswritten ist %d", bytesWritten);
-    
+        LOG_INFO("byteswritten ist %d", bytesWritten);
         if(bytesWritten == commandsize)
```
Comment (nhjschulz) [inline lib/ZumoOta/src/FlashManager.cpp:124]: blank after if

File: lib/ZumoOta/src/BootloaderCom.h [hunk 9]
```diff
@@ -193,13 +200,14 @@ class BootloaderCom
     /** 
      *Specifies the current memory/page address for flashing the firmware of the Zumo program flash.
      */
-    uint16_t m_currWriteMemAddr;
+    uint16_t static m_currWriteMemAddr;
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.h:203]: unused remove

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 10]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+    size_t m_firmwareBytesRead; /**< Number of bytes read from the firmware file.*/
+    File m_firmwareFile; /**< Handle for the firmware file.*/
+
+    /**
+     * @brief Enumeration of programming commands.
+     * This enumeration defines the various programming commands used during firmware programming.
+     * It includes commands for setting memory addresses and writing data blocks.
+     */
+   enum ProgCmds {
+    FLSPRG_SET_ADDR,    /**< Command to set the memory address. */
```
Comment (nhjschulz) [inline lib/ZumoOta/src/BootloaderCom.cpp:232]: indent wrong

**Review by nhjschulz — CHANGES_REQUESTED** [review]
Still some issues und formattings

File: lib/ZumoOta/src/BootloaderCom.cpp [hunk 11]
```diff
@@ -95,28 +105,241 @@ class IntProgCmdProvider: public CmdProvider {
...
+    virtual bool next(const CommandInfo*& cmd, const ResponseInfo*& rsp)
+    {
+        if (m_nextCmd == FLSPRG_COMPLETE) 
+        {
+            cmd = nullptr;
+            rsp = nullptr;
+
+            if (m_firmwareFile)
+            {
+                m_firmwareFile.close(); // Close the file
```
[the remark this answers is not in the document]
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:177]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 1 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:187]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 2 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:289]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 3 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:310]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 6 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:273]: Each address addresses one WORD with 2(!) bytes, this is why one page consists of 64 addresses but 128 bytes memory.

File: lib/ZumoOta/src/FileManager.h (hunk 7 above)
Comment (chezlyne) [reply lib/ZumoOta/src/FileManager.h:18]: It will be removed next time.

File: lib/ZumoOta/src/FlashManager.cpp (hunk 8 above)
Comment (chezlyne) [reply lib/ZumoOta/src/FlashManager.cpp:124]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.h (hunk 9 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.h:203]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/BootloaderCom.cpp (hunk 10 above)
Comment (chezlyne) [reply lib/ZumoOta/src/BootloaderCom.cpp:232]: It will be taken care of by the next commit.

**Pull request closed without merging**